### PR TITLE
Kafka with protobuf format

### PR DESCRIPTION
**Pull request by vitlibar**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

https://github.com/yandex/ClickHouse/issues/4710
https://github.com/yandex/ClickHouse/issues/4744

Category (leave one):
- Bug Fix

### CODE REVIEW

File: dbms/src/IO/DelimitedReadBuffer.h [hunk 1]
```diff
@@ -36,7 +36,7 @@ class DelimitedReadBuffer : public ReadBuffer
                 return false;
 
             BufferBase::set(buffer->position(), buffer->available(), 0);
-            put_delimiter = true;
+            put_delimiter = (delimiter != 0);
```
Comment (vitlibar) [inline dbms/src/IO/DelimitedReadBuffer.h]: When using protobuf messages a single zero byte means a message with all its fields set to their default values. It seems there is no format where putting a single zero byte as a row delimiter could be useful. So I've decided not to put a row delimiter if it's set to `'\0'`.

File: dbms/src/Core/SettingsCommon.cpp [hunk 2]
```diff
@@ -552,9 +546,10 @@ void SettingChar::set(char x)
 
 void SettingChar::set(const String & x)
 {
-    checkStringIsACharacter(x);
-    value = x[0];
-    changed = true;
+    if (x.size() > 1)
```
Comment (vitlibar) [inline dbms/src/Core/SettingsCommon.cpp]: Minor change to allow setting `SettingChar` to `'\0'` just specifying `''`.
For example,
```
create table simple (
  t UInt64,
  url String
) ENGINE=Kafka SETTINGS kafka_broker_list = 'localhost:9092',
                        kafka_topic_list = 'topic',
                        kafka_group_name = 'mygroup',
                        kafka_format = 'Protobuf',
                        kafka_row_delimiter = '',
                        kafka_schema = 'simple:AccessLog';
```

File: dbms/tests/integration/helpers/docker_compose_kafka.yml [hunk 3]
```diff
@@ -15,14 +15,17 @@ services:
...
     ports:
-      - "9092:9092"
+        - "9092:9092"
     environment:
-      KAFKA_ADVERTISED_LISTENERS: "PLAINTEXT://kafka1:9092"
-      KAFKA_ZOOKEEPER_CONNECT: "kafka_zookeeper:2181"
-      KAFKA_BROKER_ID: 1
-      KAFKA_LOG4J_LOGGERS: "kafka.controller=INFO,kafka.producer.async.DefaultEventHandler=INFO,state.change.logger=INFO"
-      KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR: 1
+        KAFKA_ADVERTISED_LISTENERS: INSIDE://localhost:9092,OUTSIDE://kafka1:19092
```
Comment (vitlibar) [inline dbms/tests/integration/helpers/docker_compose_kafka.yml]: These changes were made to allow the python test `test_kafka_protobuf` to publish messages to kafka.

File: dbms/programs/client/Client.cpp [hunk 4]
```diff
@@ -796,14 +796,33 @@ class Client : public Poco::Util::Application
         written_progress_chars = 0;
         written_first_block = false;
 
-        connection->forceConnected();
+        {
+            /// Temporarily apply query settings to context.
```
Comment (abyss7) [inline dbms/programs/client/Client.cpp]: Why do we need to do apply all server-side settings on the client-side here?

File: dbms/src/Core/SettingsCommon.cpp [hunk 5]
```diff
@@ -552,9 +546,10 @@ void SettingChar::set(char x)
...
+        throw Exception("A setting's value string has to be an exactly one character long", ErrorCodes::SIZE_OF_FIXED_STRING_DOESNT_MATCH);
+    char c = (x.size() == 1) ? x[0] : '\0';
```
Comment (abyss7) [inline dbms/src/Core/SettingsCommon.cpp]: How can we now distinguish the empty setting and 'zero'-char setting values?

File: dbms/programs/client/Client.cpp (hunk 4 above)
Comment (vitlibar) [reply dbms/programs/client/Client.cpp]: It doesn't seem to be trivial to distinguish at this point which setting are client-side and which are server-side. It's easier to just apply all the settings.

File: dbms/src/Core/SettingsCommon.cpp (hunk 5 above)
Comment (vitlibar) [reply dbms/src/Core/SettingsCommon.cpp]: We don't distinguish them. The following two declaration mean the same:

```
create table simple (
  t UInt64,
  url String
) ENGINE=Kafka SETTINGS kafka_broker_list = 'localhost:9092',
                        kafka_topic_list = 'topic',
                        kafka_group_name = 'mygroup',
                        kafka_format = 'Protobuf',
                        kafka_row_delimiter = '',
                        kafka_schema = 'simple:AccessLog';
```
and
```
create table simple (
  t UInt64,
  url String
) ENGINE=Kafka SETTINGS kafka_broker_list = 'localhost:9092',
                        kafka_topic_list = 'topic',
                        kafka_group_name = 'mygroup',
                        kafka_format = 'Protobuf',
                        kafka_row_delimiter = '\0',
                        kafka_schema = 'simple:AccessLog';
```

File: dbms/src/IO/DelimitedReadBuffer.h (hunk 1 above)
Comment (abyss7) [reply dbms/src/IO/DelimitedReadBuffer.h]: This change will break compatibility with previous scenarios where people may have used `\0` as delimiter.

**Pull request merged**